File: kafkakewl-metrics/src/main/scala/com/mwam/kafkakewl/metrics/services/KafkaTopicInfoCache.scala
/*
 * SPDX-FileCopyrightText: 2023 Marshall Wace <opensource@mwam.com>
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package com.mwam.kafkakewl.metrics.services

import com.mwam.kafkakewl.metrics.domain.KafkaTopicPartitionInfoExtensions.*
import com.mwam.kafkakewl.metrics.domain.{KafkaSingleTopicPartitionInfos, KafkaTopicPartitionInfoChanges}
import zio.*

class KafkaTopicInfoCache(
    private val kafkaTopicInfosRef: Ref[Map[String, KafkaSingleTopicPartitionInfos]]
) {
  def getTopics: UIO[Seq[String]] = kafkaTopicInfosRef.get.map(_.keys.toSeq.sorted)
  def getTopicPartitionInfos(topic: String): UIO[Option[KafkaSingleTopicPartitionInfos]] = kafkaTopicInfosRef.get.map(_.get(topic))
}

object KafkaTopicInfoCache {
  def live: ZLayer[KafkaTopicInfoSource, Nothing, KafkaTopicInfoCache] =
    ZLayer.scoped {
      for {
        topicChangesDequeue <- ZIO.serviceWithZIO[KafkaTopicInfoSource](_.subscribe())
        kafkaTopicInfosRef <- Ref.make(Map.empty[String, KafkaSingleTopicPartitionInfos])
        processTopicChangesFiber <- processTopicChanges(topicChangesDequeue, kafkaTopicInfosRef).forever.fork
        _ <- ZIO.addFinalizer(processTopicChangesFiber.interrupt)
      } yield KafkaTopicInfoCache(kafkaTopicInfosRef)
    }

  private def processTopicChanges(
      topicInfoChangesDequeue: Dequeue[KafkaTopicPartitionInfoChanges],
      topicPartitionInfosRef: Ref[Map[String, KafkaSingleTopicPartitionInfos]]
  ) = for {
    topicInfoChanges <- topicInfoChangesDequeue.take
    _ <- topicPartitionInfosRef.update(_.applyChanges(topicInfoChanges))
    _ <- ZIO.logInfo(
      s"applied ${topicInfoChanges.addedOrUpdated.size} added or updated topic infos and ${topicInfoChanges.removed.size} removed ones"
    )
  } yield ()
}

File: kafkakewl-metrics/src/main/kotlin/com/mwam/kafkakewl/metrics/services/KafkaTopicInfoCache.kt
/*
 * SPDX-FileCopyrightText: 2025 Marshall Wace <opensource@mwam.com>
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package com.mwam.kafkakewl.metrics.services

import com.mwam.kafkakewl.metrics.domain.KafkaSingleTopicPartitionInfos
import com.mwam.kafkakewl.metrics.domain.KafkaTopicPartitionInfoChanges
import com.mwam.kafkakewl.metrics.domain.applyChanges
import io.github.oshai.kotlinlogging.KotlinLogging
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import java.util.concurrent.atomic.AtomicReference

interface KafkaTopicInfoCache : AutoCloseable {
    fun getTopics(): List<String>
    fun getTopicPartitionInfos(topic: String): KafkaSingleTopicPartitionInfos?
}

private val logger = KotlinLogging.logger {}

class KafkaTopicInfoCacheImpl(
    source: KafkaTopicInfoSource
) : KafkaTopicInfoCache {
    private val scope = CoroutineScope(Dispatchers.Default)
    private val topicInfoChangesCollectJob: Job?
    private val topicPartitionInfosRef: AtomicReference<Map<String, KafkaSingleTopicPartitionInfos>> = AtomicReference(emptyMap())

    init {
        topicInfoChangesCollectJob = scope.launch { source.topicInfoChanges().collect { onTopicInfoChanges(it)} }
    }

    override fun getTopics(): List<String> = topicPartitionInfosRef.get().keys.sorted()
    override fun getTopicPartitionInfos(topic: String): KafkaSingleTopicPartitionInfos? = topicPartitionInfosRef.get()[topic]

    override fun close() {
        logger.info { "Closing..." }
        runBlocking {
            scope.cancel()
            topicInfoChangesCollectJob?.join()
        }
        logger.info { "Finished closing" }
    }

    private fun onTopicInfoChanges(changes: KafkaTopicPartitionInfoChanges): Unit {
        topicPartitionInfosRef.set(topicPartitionInfosRef.get().applyChanges(changes))
    }
}
